services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreRepairTask.java

```diff
@@ -0,0 +1,255 @@
+      if (store.getCurrentVersion() == 0) {
+        newUnhealthySystemStoreSet.add(store.getName());
+        continue;
+      }
+
+      // Send heartbeat to system store in all child regions.
+      long currentTimestamp = System.currentTimeMillis();
+      sendHeartbeatToSystemStore(clusterName, store.getName(), currentTimestamp);
+      systemStoreToHeartbeatTimestampMap.put(store.getName(), currentTimestamp);
+
+      // Sleep to throttle heartbeat send rate.
+      try {
+        Thread.sleep(heartbeatCheckIntervalMs);
+      } catch (InterruptedException e) {
+        LOGGER.info("Caught interrupted exception, will exit now.");
+        return;
+      }
+    }
+  }
+
+  /**
+   * This method iterates over all system stores and validate if heartbeat has been received.
+   */
+  void checkHeartbeatFromSystemStores(
+      String clusterName,
+      Set<String> newUnhealthySystemStoreSet,
+      Map<String, Long> systemStoreToHeartbeatTimestampMap) {
+    for (Map.Entry<String, Long> entry: systemStoreToHeartbeatTimestampMap.entrySet()) {
+      if (!shouldContinue(clusterName)) {
+        return;
+      }
+      if (!isHeartbeatReceivedBySystemStore(clusterName, entry.getKey(), entry.getValue())) {
+        newUnhealthySystemStoreSet.add(entry.getKey());
+      }
+    }
+  }
+
+  void updateBadSystemStoreCount(Set<String> newUnhealthySystemStoreSet) {
+    long newBadMetaSystemStoreCount = newUnhealthySystemStoreSet.stream()
+        .filter(x -> VeniceSystemStoreType.getSystemStoreType(x).equals(VeniceSystemStoreType.META_STORE))
+        .count();
+    getBadMetaStoreCount().set(newBadMetaSystemStoreCount);
+    getBadPushStatusStoreCount().set(newUnhealthySystemStoreSet.size() - newBadMetaSystemStoreCount);
+    LOGGER.info(
+        "Collected unhealthy system stores:{}. Meta system store count: {}, push status system store count: {}",
+        newUnhealthySystemStoreSet.toString(),
+        getBadMetaStoreCount().get(),
+        getBadPushStatusStoreCount().get());
+  }
+
+  AtomicBoolean getIsRunning() {
+    return isRunning;
+  }
+
+  public Map<String, ControllerClient> getControllerClientMap(String clusterName) {
+    return getParentAdmin().getVeniceHelixAdmin().getControllerClientMap(clusterName);
+  }
+
+  void sendHeartbeatToSystemStore(String clusterName, String systemStoreName, long heartbeatTimestamp) {
+    for (Map.Entry<String, ControllerClient> entry: getControllerClientMap(clusterName).entrySet()) {
+      entry.getValue().sendHeartbeatToSystemStore(systemStoreName, heartbeatTimestamp);
+    }
+  }
+
+  boolean isHeartbeatReceivedBySystemStore(
+      String clusterName,
+      String systemStoreName,
+      long expectedHeartbeatTimestamp) {
+    for (Map.Entry<String, ControllerClient> entry: getControllerClientMap(clusterName).entrySet()) {
+      long heartbeatFromChildController =
+          entry.getValue().getHeartbeatFromSystemStore(systemStoreName).getHeartbeatTimestamp();
```

Comment:
I think we should check whether the response contains any error or not.
If the response is errored, we shouldn't mark heart beat failure since it might indicate some serious issue, which we need to fix manually.
Let us also emit a metric to track error responses.

Reply:
Added to track unreachableSystemStore.